modules/BuildTools.Tasks/GenerateFileFromTemplate.cs

```diff
@@ -0,0 +1,179 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.AspNetCore.BuildTools
+{
+    /// <summary>
+    /// <para>
+    /// Generates a new file at <see cref="OutputPath"/>.
+    /// </para>
+    /// <para>
+    /// The <see cref="TemplateFile"/> can define variables for substitution using <see cref="Properties"/>.
+    /// </para>
+    /// <example>
+    /// The input file might look like this:
+    /// <code>
+    /// 2 + 2 = ${Sum}
+    /// </code>
+    /// When the task is invoked like this, it will produce "2 + 2 = 4"
+    /// <code>
+    /// &lt;GenerateFileFromTemplate Properties="Sum=4;OtherValue=123;" ... &gt;
+    /// </code>
+    /// </example>
+    /// </summary>
+#if SDK
+    public class Sdk_GenerateFileFromTemplate : Task
+#elif BuildTools
+    public class GenerateFileFromTemplate : Task
+#else
+#error This must be built either for an SDK or for BuildTools
+#endif
+    {
+        /// <summary>
+        /// The template file.
+        /// Variable syntax: ${VarName}
+        /// If your template file needs to output this format, you can escape the dollar sign with a backtick, e.g. `${NotReplaced}
+        /// </summary>
+        [Required]
+        public string TemplateFile { get; set; }
+
+        /// <summary>
+        /// The desitnation for the generated file.
+        /// </summary>
+        [Required]
+        public string OutputPath { get; set; }
+
+        /// <summary>
+        /// Key=value pairs of values
+        /// </summary>
+        [Required]
+        public string[] Properties { get; set; }
+
+        /// <summary>
+        /// The full path of the file written.
+        /// </summary>
+        [Output]
+        public string FileWrites { get; set; }
+
+        public override bool Execute()
+        {
+            var outputPath = Path.GetFullPath(OutputPath.Replace('\\', '/'));
+
+            if (!File.Exists(TemplateFile))
+            {
+                Log.LogError("File {0} does not exist", TemplateFile);
+                return false;
+            }
+
+            var values = MSBuildListSplitter.GetNamedProperties(Properties);
+            var template = File.ReadAllText(TemplateFile);
+
+            var result = Replace(template, values);
+            Directory.CreateDirectory(Path.GetDirectoryName(outputPath));
+            File.WriteAllText(outputPath, result);
+
+            FileWrites = Path.GetFullPath(outputPath);
+
+            return true;
+        }
+
+        internal string Replace(string template, IDictionary<string, string> values)
+        {
+            var sb = new StringBuilder();
+            var line = 1;
+            for (var i = 0; i < template.Length; i++)
+            {
+                var ch = template[i];
+                if (ch == '\n')
```

Comment:
The logic here is a little convoluted, any chance we can simplify this with a case statement instead? Also is the reason we are iterating through the template instead of doing a simple `.Replace` that we want to produce useful warning messages?

Reply:
Calling .Replace would be ok if we only ever had one variable. Calling it with multiple variables would iterate over the document multiple times

Reply:
Ah so it's for perf reasons.